index.html

```diff
@@ -466,21 +486,44 @@
 
     charCanvas.addEventListener('mousemove', (event) => {
       if (isDragging) {
-        dragEnd = { x: event.offsetX, y: event.offsetY };
+        const chartArea = budgetChart.chartArea;
+
+        // Obliczenie marginesu 2%
+        const marginX = (chartArea.right - chartArea.left) * 0.02;
+        const marginY = (chartArea.bottom - chartArea.top) * 0.02;
+
+        // Dopasowanie `dragEnd` do krawędzi wykresu z marginesem
+        dragEnd = {
+          x: Math.min(Math.max(event.offsetX, chartArea.left - marginX), chartArea.right + marginX),
+          y: Math.min(Math.max(event.offsetY, chartArea.top - marginY), chartArea.bottom + marginY),
+        };
+
         drawSelectionBox();
       }
     });
-
-    charCanvas.addEventListener('mouseup', () => {
+    function endDragging() {
       if (isDragging && dragEnd) {
         updateChartZoom();
+        // Ukryj suwak osi i resetuj stany
+        axisSlider.style.display = 'none';
+        axisSliderValue.style.display = 'none';
+        axisSliderLabel.style.display = 'none';
+        isDragging = false;
+        dragStart = null;
+        dragEnd = null;
+      }
+    }
+
+    charCanvas.addEventListener('mouseup', () => {
+      if (isDragging) {
+        endDragging();
+      }
+    });
+
+    charCanvas.addEventListener('mouseleave', () => {
+      if (isDragging) {
+        endDragging();

```

Comment:
It was kind of hard to operate with zoom near borders, that should fix that + avoid keeping rectangle selected without zooming while leaving the area